### PR TITLE
Render-start analytics event for documents and embeds

### PR DESCRIPTION
**Pull request by dvoytenko**
`render-start` event is emitted by documents and some embeds when it's confirmed that the rendering of the document has been started.

### CODE REVIEW

File: spec/amp-var-substitutions.md [hunk 1]
```diff
@@ -569,6 +569,20 @@ Provides the type of the last non-redirect navigation in the current browsing co
 * **amp-analytics variable**: `${navType}`
   * Example value: `1`
 
+
+#### Time since created
+
+Provides the time elapsed since the creation of this document or embed.
```
Comment (cramforce) [inline spec/amp-var-substitutions.md]: Needs unit

File: extensions/amp-analytics/amp-analytics.md [hunk 2]
```diff
@@ -398,6 +398,32 @@ As an example, the following configuration can be used to sample 50% of the requ
 },
 ```
 
+#### Embed render start trigger (`"on": "render-start"`)
+Use this configuration to fire a request when this document's or a specified embed's receive render-start signal. This
```
Comment (cramforce) [inline extensions/amp-analytics/amp-analytics.md]: There are some grammatical issues here.

I think what would really help is to start out by describing what an embed (give ad example) and a document is.

File: src/service/ampdoc-impl.js [hunk 3]
```diff
@@ -183,6 +193,16 @@ export class AmpDoc {
   constructor(win) {
     /** @public @const {!Window} */
     this.win = win;
+
+    /** @private {boolean} */
+    this.isRenderStarted_ = false;
```
Comment (cramforce) [inline src/service/ampdoc-impl.js]: Why is this not signal based?

Comment (dvoytenko) [inline src/service/ampdoc-impl.js]: Do you mean you want to apply the same signals instrument to ampdocs as well? That's probably a good idea. Will do.

Comment (cramforce) [inline src/service/ampdoc-impl.js]: Probably want some kind of `Signable` interface

File: spec/amp-var-substitutions.md (hunk 1 above)
Comment (dvoytenko) [inline spec/amp-var-substitutions.md]: Tests are coming...

Comment (dvoytenko) [inline spec/amp-var-substitutions.md]: Oh. Wait. You mean unit of time. Got it.

File: extensions/amp-ad/0.1/test/test-amp-ad-xorigin-iframe-handler.js [hunk 4]
```diff
@@ -41,6 +42,10 @@ describe('amp-ad-xorigin-iframe-handler', () => {
     adElement.isBuilt = () => {
       return true;
     };
+    signals = {};
+    adElement.signal = name => {
```
Comment (cramforce) [inline extensions/amp-ad/0.1/test/test-amp-ad-xorigin-iframe-handler.js]: Not clear why you need to do this.

File: extensions/amp-analytics/amp-analytics.md [hunk 5]
```diff
@@ -398,6 +398,33 @@ As an example, the following configuration can be used to sample 50% of the requ
 },
 ```
 
+#### Embed render start trigger (`"on": "render-start"`)
+Use this configuration to fire a request when the document or a specified embed emit `render-start` signal. This
```
Comment (cramforce) [inline extensions/amp-analytics/amp-analytics.md]: `emits`

But I'd consider switching the language to not use our internal name "signal" and just make this a plain event.

File: extensions/amp-analytics/amp-analytics.md [hunk 6]
```diff
@@ -398,6 +398,33 @@ As an example, the following configuration can be used to sample 50% of the requ
...
+```javascript
+"triggers": {
+  "renderStart": {
+    "on": "render-start",
+    "request": "render-start"
+  }
+}
+```
+
+For an embed within a document, the trigger includes a selector to specify the embed element:
```
Comment (cramforce) [inline extensions/amp-analytics/amp-analytics.md]: +1 this is not clear.

File: src/service/resources-impl.js [hunk 7]
```diff
@@ -248,6 +248,14 @@ export class Resources {
   }
 
   /**
+   * Signals that the document has been started rendering.
+   * @restricted
+   */
+  renderStarted() {
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: There's more logical coming to `Resources` based on this signal.

Comment (lannka) [inline src/service/resources-impl.js]: what kind of logic?
and how should this work in amp-inabox?

Comment (dvoytenko) [inline src/service/resources-impl.js]: I'm working on moving some scheduling here as well.

This should be called and work exactly the same with amp-inabox with one exception: we also want to postMessage `render-start` which is what you are working on :) Where exactly postMessage belongs in, I'm not sure. I was going to see where you end up putting it. But my first suggestion was to stick it right into `amp-inabox.js`. That means that we may need a callback from this method, which you can add later if that's what you'd like to do.

File: extensions/amp-analytics/amp-analytics.md (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-analytics/amp-analytics.md]: @cramforce @lannka I rewrote this. Please advise if this is better.

**Comment by dvoytenko** [comment]
Tests have been added.

**Pull request merged**